samples/17.multilingual-bot/src/main/java/com/microsoft/bot/sample/multilingual/translation/TranslationMiddleware.java

```diff
@@ -0,0 +1,122 @@
+     * @param userState State property for current language.
+     */
+    public TranslationMiddleware(MicrosoftTranslator withTranslator, UserState userState) {
+        if (withTranslator == null) {
+            throw new IllegalArgumentException("withTranslator");
+        }
+        this.translator = withTranslator;
+        if (userState == null) {
+            throw new IllegalArgumentException("userState");
+        }
+
+        this.languageStateProperty = userState.createProperty("LanguagePreference");
+    }
+
+    /**
+     * Processes an incoming activity.
+     * @param turnContext Context object containing information for a single turn of conversation with a user.
+     * @param next The delegate to call to continue the bot middleware pipeline.
+     * @return A Task representing the asynchronous operation.
+     */
+    public CompletableFuture<Void> onTurn(TurnContext turnContext, NextDelegate next) {
+        if (turnContext == null) {
+            throw new IllegalArgumentException("turnContext");
+        }
+
+        return this.shouldTranslate(turnContext).thenCompose(translate -> {
+            if (translate) {
+                if (turnContext.getActivity().getType() == ActivityTypes.MESSAGE) {
+                    return this.translator.translate(turnContext.getActivity().getText(), TranslationSettings.getDefaultLanguage()).thenApply(text -> {
+                        turnContext.getActivity().setText(text);
+                    });
+                }
+
+                return CompletableFuture.completedFuture(null);
+            }
+        }).thenCompose(task -> {
+            turnContext.onSendActivities((newContext, activities, nextSend) -> {
+                return this.languageStateProperty.get(turnContext, () -> TranslationSettings.getDefaultLanguage()).thenCompose(userLanguage -> {
+                    Boolean shouldTranslate = !userLanguage.equals(TranslationSettings.getDefaultLanguage());
+
+                    // Translate messages sent to the user to user language
+                    if (shouldTranslate) {
+                        ArrayList<CompletableFuture<Void>> tasks = new ArrayList<CompletableFuture<Void>>();
+                        for (Activity activity : activities.stream().filter(a -> a.getType() == ActivityTypes.MESSAGE).collect(Collectors.toList())) {
+                            tasks.add(this.translateMessageActivity(activity, userLanguage));
+                        }
+
+                        if (Arrays.asList(tasks).isEmpty()) {
+                            CompletableFuture.allOf(tasks.toArray(new CompletableFuture[tasks.size()]));
+                        }
+
+                        return nextSend.get();
+                    }
+
+                    return CompletableFuture.completedFuture(null);
+                });
+            });
+
+            turnContext.onUpdateActivity((newContext, activity, nextUpdate) -> {
```

Comment:
This `onUpdateActivity` method shouldn't be added as `thenCompose` of the second `CompletableFuture` `onSendActivities`?

Reply:
Not necessary, both can be executed in parallel, there is no dependency. Also i checked the [InterceptionMiddleware](https://github.com/microsoft/botbuilder-java/blob/main/libraries/bot-builder/src/main/java/com/microsoft/bot/builder/inspection/InterceptionMiddleware.java) which is similar.

Both execution can be executed in parallel

Reply:
Even with the return of above (line 54)? is this code block will execute anyway?

Reply:
Think that the line 54 is inside another `thenCompose`, as soon as that finishes, it will go to the `onSend` + `onUpdate` in parallel